pkg/config/lifecycle_provider.go

```diff
@@ -0,0 +1,124 @@
+package config
+
+import (
+	"sync/atomic"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	LifecycleConfigName = "lifecycle-image"
+	LifecycleConfigKey  = "image"
+)
+
+type RegistryClient interface {
+	Fetch(keychain authn.Keychain, repoName string) (v1.Image, string, error)
+}
+
+type lifecycleData struct {
+	image v1.Image
+	err   error
+}
+
+type LifecycleProvider struct {
+	RegistryClient RegistryClient
+	Keychain       authn.Keychain
+	lifecycleData  atomic.Value
+	handlers       []func()
+}
+
+func NewLifecycleProvider(lifecycleImageRef string, client RegistryClient, keychain authn.Keychain) *LifecycleProvider {
+	p := &LifecycleProvider{
+		RegistryClient: client,
+		Keychain:       keychain,
+	}
+
+	data := &lifecycleData{}
+
+	data.image, data.err = p.fetchImage(lifecycleImageRef)
+
+	p.lifecycleData.Store(data)
+	return p
+}
+
+func (l *LifecycleProvider) UpdateImage(cm *corev1.ConfigMap) {
```

Comment:
Here's the best that I could come up with. Hopefully there's a good idea in there somewhere:

```suggestion
func (l *LifecycleProvider) UpdateImage(cm *corev1.ConfigMap) {
	data, shouldCallCallBacks := l.updateImage(cm)
	if shouldCallCallBacks {
		l.callCallBacks()
	}

	l.lifecycleData.Store(data)
}

func (l *LifecycleProvider) updateImage(cm *corev1.ConfigMap) (*lifecycleData, bool) {
	data := &lifecycleData{}

	imageRef, ok := cm.Data[LifecycleConfigKey]
	if !ok {
		data.err = errors.Errorf("%s config invalid", LifecycleConfigName)
		return data, true
	}

	l.fetchImage(imageRef, data)
	if data.err != nil {
		return data, true
	}

	// Don't care if old image errored
	oldImg, _ := l.GetImage()

	var isNewImg bool
	isNewImg, data.err = isNewImage(oldImg, data.image)
	return data, isNewImg
}

func (l *LifecycleProvider) fetchImage(imageRef string, v *lifecycleData) {
	img, _, err := l.RegistryClient.Fetch(l.Keychain, imageRef)
	if err != nil {
		v.err = errors.Wrap(err, "failed to fetch lifecycle image")
		return
	}

	v.image = img
}
```

Reply:
I like this one!